src/backend/distributed/utils/background_jobs.c

```diff
@@ -385,292 +629,347 @@ CitusBackgroundTaskQueueMonitorMain(Datum arg)
 		PopActiveSnapshot();
 		CommitTransactionCommand();
 
-		MemoryContextSwitchTo(perTaskContext);
+		/* get running task entries from hash table */
+		List *runningTaskEntries = GetRunningTaskEntries(backgroundExecutorHandles);
+		hasAnyTask = list_length(runningTaskEntries) > 0;
 
-		bool hadError = false;
-		StringInfoData message = { 0 };
-		initStringInfo(&message);
+		/* useful to sleep if all tasks ewouldblock on current iteration */
+		bool allWouldBlock = true;
 
+		/* iterate over all handle entries and monitor each task's output */
+		BackgroundExecutorHashEntry *handleEntry = NULL;
+		foreach_ptr(handleEntry, runningTaskEntries)
 		{
-			shm_toc *toc = shm_toc_attach(CITUS_BACKGROUND_TASK_MAGIC,
-										  dsm_segment_address(seg));
-			shm_mq *mq = shm_toc_lookup(toc, CITUS_BACKGROUND_TASK_KEY_QUEUE, false);
-			shm_mq_handle *responseq = shm_mq_attach(mq, seg, NULL);
+			/* execution state for current task in FSM */
+			BackgroundMonitorExecutionStates monitorExecutionState = ExecutionStarted;
 
-			/*
-			 * We consume the complete shm_mq here as ConsumeTaskWorkerOutput loops till
-			 * it reaches a SHM_MQ_DETACHED response.
-			 */
-			ConsumeTaskWorkerOutput(responseq, &message, &hadError);
+			/* useful to store cancelling status for current task in FSM */
+			bool taskCancelledConcurrently = false;
 
-			shm_mq_detach(responseq);
-		}
-
-		StartTransactionCommand();
-		PushActiveSnapshot(GetTransactionSnapshot());
+			/* currently running task used in FSM */
+			BackgroundTask *task = NULL;
 
-		/*
-		 * Same as before, we need to lock pg_dist_background_task in a way where we can
-		 * check if there had been a concurrent cancel.
-		 */
-		LockRelationOid(DistBackgroundTaskRelationId(), ExclusiveLock);
+			/* start transaction for monitoring current task */
+			StartTransactionCommand();
+			PushActiveSnapshot(GetTransactionSnapshot());
 
-		oldContext = MemoryContextSwitchTo(perTaskContext);
-		task = GetBackgroundTaskByTaskId(task->taskid);
-		MemoryContextSwitchTo(oldContext);
+			/* tracks execution status for current task execution. */
+			bool taskWouldBlock = false;
+			bool taskEnded = false;
 
-		if (!task || task->status == BACKGROUND_TASK_STATUS_CANCELLING ||
-			task->status == BACKGROUND_TASK_STATUS_CANCELLED)
-		{
 			/*
-			 * A concurrent cancel has happened or the task has disappeared, we are not
-			 * retrying or changing state, we will only reflect the message onto the task
-			 * and for completeness we update the job aswell, this should be a no-op
-			 *
-			 * We still need to release the shared memory and xact before looping
+			 * starting from initial FSM state (ExecutionStarted), loop until
+			 * current task has no output (EWOULDBLOCK) or it is ended.
+			 * (suceeded, failed, or cancelled)
 			 */
+			while (!taskWouldBlock && !taskEnded)
+			{
+				switch (monitorExecutionState)
+				{
+					case ExecutionStarted:
```

Comment:
I don't think having this be a state machine adds anything to the readability. The flow seems completely linear:

1. ExecutionStarted
2. TaskConcurrentCancelCheck
3. TryConsumeTaskWorker
4. if (error) TaskHadError
5. if (ended) TaskEnded